ads/google/core-shared.js

```diff
@@ -0,0 +1,482 @@
+
+/**
+ * The layout - aspect ratio map to calculate the size of each content
+ * recommendation.
+ * image_stacked: https://screenshot.googleplex.com/74S09gFO82b
+ * image_sidebyside: https://screenshot.googleplex.com/FUgDSDvwcVo
+ * image_card_stacked: https://screenshot.googleplex.com/vedjTonVaDT
+ * image_card_sidebyside: https://screenshot.googleplex.com/v3qOZY61tFm
+ * text: https://screenshot.googleplex.com/taeRQn7DUhq
+ * text_card: https://screenshot.googleplex.com/ur45m96Tv0D
+ * @const {!Object<!LayoutType, number>}
+ */
+const LAYOUT_ASPECT_RATIO_MAP = {
+  [LayoutType.IMAGE_STACKED]: 1 / 1.91,
+  [LayoutType.IMAGE_SIDEBYSIDE]: 1 / 3.82,
+  [LayoutType.MOBILE_BANNER_IMAGE_SIDEBYSIDE]: 1 / 3.82,
+  [LayoutType.PUB_CONTROL_IMAGE_STACKED]: 1 / 1.91,
+  [LayoutType.PUB_CONTROL_IMAGE_SIDEBYSIDE]: 1 / 3.82,
+  [LayoutType.PUB_CONTROL_IMAGE_CARD_STACKED]: 1 / 1.91,
+  [LayoutType.PUB_CONTROL_IMAGE_CARD_SIDEBYSIDE]: 1 / 3.74,
+  [LayoutType.PUB_CONTROL_TEXT]: 0,
+  [LayoutType.PUB_CONTROL_TEXT_CARD]: 0,
+};
+
+/**
+ * The layout - height map to evaluate the height of title + url. Notice, this
+ * mainly works only for stacked mode. In sidebyside mode, this height is
+ * decided by the height of image. It equals to:
+ * FontSize * LineHeight * NumTitle + Padding * 2 + UrlBoxHeight.
+ * image_stacked: https://screenshot.googleplex.com/74S09gFO82b
+ * image_card_stacked: https://screenshot.googleplex.com/vedjTonVaDT
+ * @const {!Object<!LayoutType, number>}
+ */
+const LAYOUT_TEXT_HEIGHT_MAP = {
+  [LayoutType.IMAGE_STACKED]: 80,
+  [LayoutType.IMAGE_SIDEBYSIDE]: 0,
+  [LayoutType.MOBILE_BANNER_IMAGE_SIDEBYSIDE]: 0,
+  [LayoutType.PUB_CONTROL_IMAGE_STACKED]: 80,
+  [LayoutType.PUB_CONTROL_IMAGE_SIDEBYSIDE]: 0,
+  [LayoutType.PUB_CONTROL_IMAGE_CARD_STACKED]: 85,
+  [LayoutType.PUB_CONTROL_IMAGE_CARD_SIDEBYSIDE]: 0,
+  [LayoutType.PUB_CONTROL_TEXT]: 80,
+  [LayoutType.PUB_CONTROL_TEXT_CARD]: 80,
+};
+
+/**
+ * The layout - minimal width map for pub control UIs. We will adjust column
+ * numbers according to minimal width.
+ * @const {!Object<!LayoutType, number>}
+ */
+const LAYOUT_AD_WIDTH_MAP = {
+  [LayoutType.PUB_CONTROL_IMAGE_STACKED]: 100,
+  [LayoutType.PUB_CONTROL_IMAGE_SIDEBYSIDE]: 200,
+  [LayoutType.PUB_CONTROL_IMAGE_CARD_STACKED]: 150,
+  [LayoutType.PUB_CONTROL_IMAGE_CARD_SIDEBYSIDE]: 250,
+  [LayoutType.PUB_CONTROL_TEXT]: 100,
+  [LayoutType.PUB_CONTROL_TEXT_CARD]: 150,
+};
+
+const PUB_CONTROL_LAYOUT_PREFIX = 'pub_control_';
+
+const PUB_CONTROL_EXAMPLE = '\n ' +
+  'data-matched-content-rows-num=\"4,2\"\n' +
+  'data-matched-content-columns-num=\"1,6\"\n' +
+  'data-matched-content-ui-type=\"image_stacked,image_card_sidebyside\"';
+
+/**
+ * Configuration of content recommendation unit for current slot. This is the
+ * result of running CoRe responsive logic and values from this config
+ * will be used in ad request.
+ * @record
+ */
+class CoReConfig { // eslint-disable-line no-unused-vars
```

Comment:
Does a TypeDef work here? Is a class needed?

Reply:
Generally styleguide recommends records over typedefs: https://google.github.io/styleguide/jsguide.html#jsdoc-enum-and-typedef-comments